### PR TITLE
Add new HOLETYPES to TASTy format 

### PR DESCRIPTION
**Pull request by nicolasstucki**
This is a new encoding of HOLE that differentiates between type and
term arguments of the hole.

```
Term      = ...
            HOLE Length idx_Nat tpe_Type HoleTypes? arg_Tree*
                -- Splice hole with index `idx`, the type of the hole `tpe`,
                   type arguments of the hole `targ`s and term arguments
                   of the hole `arg`s
HoleTypes = HOLETYPES Length targ_Type*
```

We will only have hole `targs` if there is a type defined in a quote and used in a nested quote. Most of the time we do not have those types and therefore no overhead in the encoding compared to before this change.

Based on #17144

### CODE REVIEW

File: tasty/src/dotty/tools/tasty/TastyFormat.scala [hunk 1]
```diff
@@ -122,8 +122,9 @@ Standard-Section: "ASTs" TopLevelStat*
                   MATCHtpt       Length bound_Term? sel_Term CaseDef*              -- sel match { CaseDef } where `bound` is optional upper bound of all rhs
                   BYNAMEtpt             underlying_Term                            -- => underlying
                   SHAREDterm            term_ASTRef                                -- Link to previously serialized term
-                  HOLE           Length idx_Nat tpe_Type arg_Tree*                 -- Splice hole with index `idx`, the type of the hole `tpe`, type and term arguments of the hole `arg`s
-
+                  HOLE           Length idx_Nat tpe_Type (HOLETYPES Length targ_Type*)? arg_Tree*
+                                                                                   -- Splice hole with index `idx`, the type of the hole `tpe`, type arguments of the hole `targ`s and term arguments of the hole `arg`s
+                                                                                      -- Note: From 3.0 to 3.3 `args` could contain type arguments as well. This is no longer the case.
```
Comment (smarter) [inline tasty/src/dotty/tools/tasty/TastyFormat.scala:127]: To be clear, args when you say "type arguments" here you mean TypeTrees, as opposed to the use of "type arguments" in the previous sentence where they are Types?

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 2]
```diff
@@ -1506,6 +1500,18 @@ class TreeUnpickler(reader: TastyReader,
...
+      val idx = readNat()
+      val tpe = readType()
+      val targs =
+        if nextByte == HOLETYPES then
+          readByte()
+          val typesEnd = readEnd()
+          until(typesEnd)(readType()).map(TypeTree(_))
+        else Nil
+      val args = until(end)(readTerm())
+      TastyQuoteHole(true, idx, targs ::: args, TypeTree(tpe)).withType(tpe)
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1513]: Maybe not. It is useful to have it in the TASTy HOLE. But we might be able to remove it from here.

[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1513]: If we remove it, we don't have a way to retype the tree. I do not have another way to infer the type of hole. This is an issue when Ychecking the tree.

[the remark this answers is not in the document]
Comment (smarter) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1512]: The TreeChecker uses `promote` to retype tree with their existing type.

**Pull request closed without merging**